### PR TITLE
adding some composition options for modals

### LINKED ISSUE #12899: Edit permission not showing save button and not scrollale on 14inch laptop
### Bug description

image

### Steps to reproduce

goto https://gitpod.io/integrations, click edit permissions.

### Workspace affected

_No response_

### Expected behavior

_No response_

### Example repository

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by selfcontained**
## Description
The intention here is to fix the linked issue below so modals are more responsive, and the body of a modal can shrink and scroll when needed. To accomplish that, I needed to add a few css classes to the modal container and body wrapper. In order to allow us to have these sections of a modal consistently rendered, I created a few sub-components so we can compose our Modals a bit as well. `<Modal/>`s can still use the `title` prop and `buttons` prop, which will get wrapped automatically like they were. It might be a good idea to unify how we render modals though at some point, and always compose them like below, but not an urgent change we have to make. These changes should provide a responsive modal w/ scrollable body for any pre-existing modal that was passing in a `title` prop as well.

I've also moved the `classnames` dependency from devDeps to normal deps as I wanted to use it for some conditional class names. I'm happy to adjust that back though if we explicitly don't want to use `classnames` outside of tests though.

```
<Modal ...>
  <ModalHeader ...>
  <ModalContent ...>
  <ModalFooter ...>
</Modal>
```

*We can iterate on this, and add built-in behavior to support pinned error messages on a modal that I saw in at least one place (when adding a new integration). This would allow that style of modal to also render responsively for smaller screen sizes.*

## Related Issue(s)

Fixes #12899, #10747

## How to test
Navigate to [Settings/Integrations](https://bmh-scroll4fb8edc1e5.preview.gitpod-dev.com/integrations) and edit one of the integrations to open the modal. 
* Resize your browser window to be smaller than the height of the modal, and you should see it shrink, while being able to scroll the inner contents of the modal.
image
* I've applied the same updates to the Environment Variables modal.

## Release Notes

```release-note
Fixes an issue with modals not displaying properly on smaller screens.
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@selfcontained: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/dashboard/src/components/Modal.tsx [hunk 1]
```diff
@@ -4,7 +4,8 @@
  * See License-AGPL.txt in the project root for license information.
  */
 
-import { useEffect } from "react";
+import { ReactNode, useEffect } from "react";
+import cn from "classnames";
```
Comment (AlexTugarev) [inline components/dashboard/src/components/Modal.tsx:8]: That makes sense. Would you make a drive-by fix the [package.json](https://github.com/gitpod-io/gitpod/blob/8dd290cf0be4baa2cd6df5e87729f78c782e9722/components/dashboard/package.json#L46) to make it a runtime dependency, please? 🙏🏻

File: components/dashboard/src/components/Modal.tsx [hunk 2]
```diff
@@ -14,8 +15,8 @@ export default function Modal(props: {
     specify?: string;
     title?: string;
     hideDivider?: boolean;
-    buttons?: React.ReactChild[] | React.ReactChild;
-    children: React.ReactChild[] | React.ReactChild;
+    buttons?: ReactNode;
```
Comment (selfcontained) [inline components/dashboard/src/components/Modal.tsx:18]: `ReactNode` type encompasses arrays or single children, but also allows for `null`.

File: components/dashboard/src/components/Modal.tsx [hunk 3]
```diff
@@ -75,10 +76,10 @@ export default function Modal(props: {
         <div className="fixed top-0 left-0 bg-black bg-opacity-70 z-50 w-screen h-screen">
             <div className="w-screen h-screen align-middle" style={{ display: "table-cell" }}>
                 <div
-                    className={
-                        "relative bg-white dark:bg-gray-900 border border-gray-200 dark:border-gray-800 rounded-xl p-6 max-w-lg mx-auto text-left " +
-                        (props.className || "")
-                    }
+                    className={cn(
+                        "flex flex-col max-h-screen relative bg-white dark:bg-gray-900 border border-gray-200 dark:border-gray-800 rounded-xl p-6 max-w-lg mx-auto text-left ",
```
Comment (selfcontained) [inline components/dashboard/src/components/Modal.tsx:80]: this allows the modal container to adjust height, while being height constrained to the screen viewport

**Comment by easyCZ** [comment]
When testing with Brave, on a 4k screen, the modal renders what looks like scroll bars, without actually being scrollable.

Screenshot 2022-12-02 at 16 36 12
Screenshot 2022-12-02 at 16 35 41

**Comment by selfcontained** [comment]
> 

Good catch - I think this is from not using an `overflow: auto` style (browsers should hide scrollbars unless needed). I'ved updated that scrollable body to use a tailwind class that should remove the scrollbars unless needed (also it should only apply a vertical scrollbar if needed and no horizontal).

File: components/dashboard/src/components/Modal.tsx (hunk 1 above)
Comment (selfcontained) [reply components/dashboard/src/components/Modal.tsx:8]: moved it from devDeps to normal deps - is that what you were referring to? Thanks for catching that, I didn't realize it was just in devDeps)

**Review by easyCZ — APPROVED** [review]
LGTM

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/components/Modal.tsx (hunk 3 above)
Comment (gtsiolis) [inline components/dashboard/src/components/Modal.tsx:80]: **thought**: This will also resolve https://github.com/gitpod-io/gitpod/issues/10747. I've updated the PR description to also include that issue.

File: components/dashboard/src/components/Modal.tsx [hunk 4]
```diff
@@ -93,16 +94,9 @@ export default function Modal(props: {
...
-                                className={
-                                    "border-gray-200 dark:border-gray-800 -mx-6 px-6 " +
-                                    (props.hideDivider ? "" : "border-t border-b mt-2 py-4")
-                                }
-                            >
-                                {props.children}
-                            </div>
-                            <div className="flex justify-end mt-6 space-x-2">{props.buttons}</div>
+                            <ModalHeader>{props.title}</ModalHeader>
+                            <ModalBody hideDivider={props.hideDivider}>{props.children}</ModalBody>
```
Comment (gtsiolis) [inline components/dashboard/src/components/Modal.tsx:98]: **fyi**: Not sure where to add this comment, but one modal I'd love to improve after thes changes is the feedback modal. Cross-posting follow up issue (https://github.com/gitpod-io/gitpod/issues/10307) for visibility.

**Review by gtsiolis — APPROVED** [review]
Hey @selfcontained! Welcome to the team, and thanks for opening this! ✨ 

Left a few non-blocking comments below that we can split into separate follow-up issues.

Some thoughts on the UX changes of this PR:
1. The responsive design of the dashboard is being tracked in https://github.com/gitpod-io/gitpod/issues/4050. 💡 
2. Improvements like creating more flexible and compact components like https://github.com/gitpod-io/gitpod/issues/13116 could help in some cases, see early designs in https://github.com/gitpod-io/gitpod/issues/10747#issuecomment-1191506907. 💡 
3. Generally, we could aim for making the viewport scrollable when the modal content extends below the viewport. 💭 
4. But making the modal contents scroll here sounds like a good, boring (simple) MVC solution here and will also resolve issues like https://github.com/gitpod-io/gitpod/issues/10747. 🛹 
5. If scrolling is needed, this may be a good indicator to consider opening the content in a new page instead. So far, the modal component and pattern has suited well our needs, and we just started experimenting with opening content in new pages in the new access tokens product are, see `/new` and `/edit` flows in [`/tokens`](https://gitpod.io/tokens). 💡 

Since code changes have been approved from two WebApp team members, let's take a bias for action, merge this, and open any follow-up issues if needed. ✔️
